### PR TITLE
Pm 33 add password encryption

### PR DESCRIPTION
**Pull request by xHoudek**
Passwords are hashed in the backend to avoid opening up an API endpoint where an adversary could get a user's (hashed) password from their email. Downside to this is that an unencrypted password is still sent in the body during the login request. I think this is okay because it's still a "potential" password at that point, not confirmed to be correct. And if someone is logging your network traffic, you probably have bigger problems haha

This does break all premade user and shelter accounts as those passwords are hashed. Not really sure what the best way around that is. I think we may need to create our own users and shelters when setting up the DB and deploying. Another option would be to do a check on the database password to see if it is "special" -- a purposely non-hashed password -- and just login with plaintext instead of checking the hash. Please let me know your thoughts on that!

### CODE REVIEW

File: src/controllers/usersController.js [hunk 1]
```diff
@@ -133,25 +135,37 @@ const loginUser = async (req, res, next) => {
   if (success) {
     const { email, password } = req.body;
 
-    await verifyLoginCredentials(req.app.get('db'), email, password)
+    await getHashedPasswordFromEmail(req.app.get('db'), email)
```
Comment (asching7108) [inline src/controllers/usersController.js]: Can we combine verifyLoginCredentials() and getHashedPasswordFromEmail() methods into one and rename it to something like getLoginCredentials()? Cause for one the verifyLoginCredentials() method doesn't really do any verification. Secondly, the two methods make a call to the same table with the same query, just to get different fields. It would look cleaner if it's only one method, and you get both UserID and Password from the User table at once.

File: src/controllers/usersController.js [hunk 2]
```diff
@@ -133,25 +135,37 @@ const loginUser = async (req, res, next) => {
...
-          res.status(401).send('Unauthorized');
+        if (isCorrectPassword(dbResponse[0].Password, password)) {
+          verifyLoginCredentials(req.app.get('db'), email)
+            .then((dbResponse) => {
+
+              // Verify email and password
+              if (dbResponse.length !== 1) {
+                // There should only ever be one user with the same email and pw
+                res.status(401).send('Unauthorized');
+              } else {
```
Comment (asching7108) [inline src/controllers/usersController.js]: The isCorrectPassword() check can go here after we verify that the email exists

File: src/controllers/sheltersController.js [hunk 3]
```diff
@@ -150,25 +151,37 @@ const loginShelter = async (req, res, next) => {
   if (success) {
     const { email, password } = req.body;
 
-    await verifyShelterLoginCredentials(req.app.get('db'), email, password)
+    await getHashedPasswordFromEmail(req.app.get('db'), email)
```
Comment (asching7108) [inline src/controllers/sheltersController.js]: Same suggestion as for the usersController

**Review by asching7108 — COMMENTED** [review]
I think it's a good choice to do the encryption in the backend! Regarding the premade accounts, I would suggest changing the passwords to hashed passwords in the sampleDataQueries.sql file. We can just use an online bcrypt password generator like https://www.browserling.com/tools/bcrypt. I personally don't think we need a special rule for certain accounts.

The code looks great, I just have one suggestion made in the comments.

File: src/controllers/usersController.js (hunk 1 above)
Comment (xHoudek) [reply src/controllers/usersController.js]: Yeah good point, I'll make that change!

**Comment by xHoudek** [comment]
Honestly didn't know I could do that with the website you linked. Super cool! Addressed all of the suggested changes

File: database/sampleDataQueries.sql [hunk 4]
```diff
@@ -32,14 +32,14 @@ VALUES
...
 
 INSERT INTO PetBreed (TypeOfAnimal, Breed)
 VALUES
   ('Cat', 'American ShortHair'),
   ('Cat', 'Domestic ShortHair'),
   ('Cat', 'Other'),
   ('Dog', 'Golden Retriever'),
-  ('Dog', 'Mixed'),
-  ('Dog', 'Other'),
+  ('Dog', 'German Shepherd'),
```
Comment (xHoudek) [inline database/sampleDataQueries.sql]: I have two GSDs, so I couldn't resist adding this when I saw it! If you want me to make this change in a different PR, just let me know!

**Review by asching7108 — APPROVED** [review]
